builtin/credential/ldap/backend.go

```diff
@@ -193,61 +193,61 @@ func getBindDN(cfg *ConfigEntry, c *ldap.Conn, username string) (string, *logica
 	return bindDN, nil
 }
 
-func getUserDN(cfg *ConfigEntry,c *ldap.Conn, binddn string) (string , *logical.Response) {
+func getUserDN(cfg *ConfigEntry,c *ldap.Conn, bindDN string) (string , error) {
 	userDN := ""
 	if cfg.UPNDomain != "" {
 		// Find the distinguished name for the user if userPrincipalName used for login
 		result, err := c.Search(&ldap.SearchRequest{
 			BaseDN: cfg.UserDN,
 			Scope:  2, // subtree
-			Filter: fmt.Sprintf("(userPrincipalName=%s)", ldap.EscapeFilter(binddn)),
+			Filter: fmt.Sprintf("(userPrincipalName=%s)", ldap.EscapeFilter(bindDN)),
 		})
 		if err != nil {
-			return userDN, logical.ErrorResponse(fmt.Sprintf("LDAP search failed for detecting user: %v", err))
+			return userDN, fmt.Errorf("LDAP search failed for detecting user: %v", err)
 		}
 		for _, e := range result.Entries {
 			userDN = e.DN
 		}
 	} else {
-		userDN = binddn
+		userDN = bindDN
 	}
 
 	return userDN, nil
 }
 
-func getLdapGroups(cfg *ConfigEntry, c *ldap.Conn, userdn string, username string) ([]string, *logical.Response) {
+func getLdapGroups(cfg *ConfigEntry, c *ldap.Conn, userDN string, username string) ([]string, error) {
 	// retrieve the groups in a string/bool map as a structure to avoid duplicates inside
 	ldapMap := make(map[string]bool)
 	// Fetch the optional memberOf property values on the user object
 	// This is the most common method used in Active Directory setup to retrieve the groups
 	result, err := c.Search(&ldap.SearchRequest{
-		BaseDN: userdn,
-		Scope:  0,        // base scope to fetch only the userdn
-		Filter: "(cn=*)", // bogus filter, required to fetch the userdn
+		BaseDN: userDN,
+		Scope:  0,        // base scope to fetch only the userDN
+		Filter: "(cn=*)", // bogus filter, required to fetch the CN from userDN
 		Attributes: []string{
 			"memberOf",
 		},
 	})
+	// this check remains in case something happens with the ldap query or connection
 	if err != nil {
-		return nil, logical.ErrorResponse(fmt.Sprintf("LDAP fetch of distinguishedName=%s failed: %v", userdn, err))
-	}
-	if len(result.Entries) != 1 {
-		return nil, logical.ErrorResponse("LDAP search for binddn 0 or not unique")
+		return nil, fmt.Errorf("LDAP fetch of distinguishedName=%s failed: %v", userDN, err)
 	}
+	// if there are more than one entry, we consider the results irrelevant and ignore them
```

Comment:
But your reasoning is right. If I look into the getUserDN function when upnDomain is not empty I set as the found userDN the DN of the last entry from the search results. And there should be only one entry here. I think I should add a check for uniqueness here also like in the bindDN discovery.
That's what is happening when you rely on the old code as a guideline.